File: blast/blast-stress-solver/src/tests/scenePack.conformance.test.ts
/**
 * Cross-runtime ScenePack conformance — TypeScript side.
 *
 * The ScenePack JSON is the contract that lets the same structure run under
 * TS/Rapier, Rust/Rapier and C++/PhysX so their APIs, behavior and performance
 * can be compared without the structure itself being a variable. That only
 * holds if all three loaders interpret the file identically.
 *
 * All three load the SAME fixture and assert the SAME digest. If a loader
 * drifts — a silently ignored field, an off-by-one index, a misread unit —
 * that runtime's own suite fails.
 *
 * The digest pins interpretation of the ASSET, not simulation results: Rapier
 * and PhysX legitimately produce different trajectories from identical input.
 *
 * See SCENE_PACK_FORMAT.md ("Conformance") and the sibling tests:
 *   demos/blast-stress-demo/tests/scene_pack_conformance_test.cpp
 *   blast/blast-stress-demo-rs/tests/scene_pack_conformance.rs
 */
import { describe, expect, it } from 'vitest';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

import { parseScenePackJson } from '../rapier/scenePackLoader';

const here = path.dirname(fileURLToPath(import.meta.url));
const conformanceDir = path.resolve(here, '../../assets/conformance');
const fixturePath = path.join(conformanceDir, 'structure-conformance-v2.json');
const goldenPath = path.join(conformanceDir, 'structure-conformance-v2.digest.json');

type Digest = {
  version: number;
  nodeCount: number;
  bondCount: number;
  materialCount: number;
  materialNames: string[];
  supportNodeCount: number;
  totalMassKg: number;
  totalBondAreaM2: number;
  bondsPerMaterial: number[];
  bondsPerJointClass: Record<string, number>;
  gravity: number;
  contactForceScale: number;
};

describe('ScenePack cross-runtime conformance', () => {
  const golden: Digest = JSON.parse(readFileSync(goldenPath, 'utf8'));
  const pack = parseScenePackJson(readFileSync(fixturePath, 'utf8'));

  it('matches the golden digest', () => {
    const bondsPerMaterial = golden.bondsPerMaterial.map(() => 0);
    for (const m of pack.bondMaterials) bondsPerMaterial[m] += 1;

    const bondsPerJointClass: Record<string, number> = {};
    for (const b of pack.scenario.bonds) {
      const key = [pack.nodeTypes[b.node0], pack.nodeTypes[b.node1]].sort().join('~');
      bondsPerJointClass[key] = (bondsPerJointClass[key] ?? 0) + 1;
    }

    const digest: Digest = {
      version: pack.version,
      nodeCount: pack.scenario.nodes.length,
      bondCount: pack.scenario.bonds.length,
      materialCount: pack.materials.length,
      materialNames: pack.materials.map((m) => m.name),
      supportNodeCount: pack.scenario.nodes.filter((n) => n.mass === 0).length,
      totalMassKg: round6(pack.scenario.nodes.reduce((sum, n) => sum + n.mass, 0)),
      totalBondAreaM2: round6(pack.scenario.bonds.reduce((sum, b) => sum + b.area, 0)),
      bondsPerMaterial,
      bondsPerJointClass: sortKeys(bondsPerJointClass),
      gravity: pack.defaults.gravity,
      contactForceScale: pack.defaults.physics.contactForceScale,
    };

    expect(digest).toEqual({ ...golden, bondsPerJointClass: sortKeys(golden.bondsPerJointClass) });
  });

  it('routes material indices to the right bonds, not just the right counts', () => {
    // Footings use the frame default; the facade clips are drywall-track.
    expect(pack.materials[pack.bondMaterials[0]].name).toBe('reinforced-concrete');
    expect(pack.materials[pack.bondMaterials[4]].name).toBe('drywall-track');
  });

  it('treats an omitted `m` as material 0 rather than unset', () => {
    // Bonds 0 and 1 carry no `m` field at all in the fixture.
    expect(pack.bondMaterials[0]).toBe(0);
    expect(pack.bondMaterials[1]).toBe(0);
  });

  it('rejects a v2 pack with no materials', () => {
    const broken = JSON.parse(readFileSync(fixturePath, 'utf8'));
    delete broken.defaults.solver.materials;
    expect(() => parseScenePackJson(JSON.stringify(broken))).toThrow(/materials is required/);
  });

  it('rejects an out-of-range bond material rather than clamping it', () => {
    // A silent clamp to material 0 would turn an authoring typo into a
    // mysteriously strong joint — the exact class of bug this format exists
    // to make impossible.
    const broken = JSON.parse(readFileSync(fixturePath, 'utf8'));
    broken.scenario.bonds[2].m = 99;
    expect(() => parseScenePackJson(JSON.stringify(broken))).toThrow(/references material 99/);
  });

  it('still loads v1 packs by synthesizing a one-entry material table', () => {
    const v1 = JSON.parse(readFileSync(fixturePath, 'utf8'));
    v1.version = 1;
    v1.defaults.solver.limits = v1.defaults.solver.materials[1];
    delete v1.defaults.solver.materials;
    for (const b of v1.scenario.bonds) delete b.m;

    const loaded = parseScenePackJson(JSON.stringify(v1));
    expect(loaded.version).toBe(1);
    expect(loaded.materials).toHaveLength(1);
    expect(loaded.bondMaterials.every((m) => m === 0)).toBe(true);
    expect(loaded.defaults.solverSettings?.compressionElasticLimit).toBe(12e6);
  });
});

function round6(value: number): number {
  return Math.round(value * 1e6) / 1e6;
}

function sortKeys(record: Record<string, number>): Record<string, number> {
  return Object.fromEntries(Object.entries(record).sort(([a], [b]) => a.localeCompare(b)));
}

File: demos/blast-stress-demo/tests/scene_pack_conformance_test.cpp
// Copyright (c) 2026 NVIDIA Corporation. All rights reserved.
//
// Cross-runtime ScenePack conformance — C++ side.
//
// The ScenePack JSON is the contract that lets the same structure run under
// TS/Rapier, Rust/Rapier and C++/PhysX so their APIs, behavior and performance
// can be compared without the structure itself being a variable. That only
// holds if all three loaders interpret the file identically. They previously
// validated `version == 1` and each read a different subset of fields, so any
// divergence — a silently ignored field, an off-by-one index, a misread unit —
// was invisible.
//
// Each runtime loads the SAME fixture and computes the SAME digest, asserted
// against one golden file. If a loader drifts, that runtime's own suite fails.
//
// The digest pins interpretation of the ASSET, not simulation results: Rapier
// and PhysX legitimately produce different trajectories from identical input.
//
// See SCENE_PACK_FORMAT.md ("Conformance") and the sibling tests:
//   blast/blast-stress-solver/src/tests/scenePack.conformance.test.ts
//   blast/blast-stress-demo-rs/tests/scene_pack_conformance.rs

#include "../scene_pack.h"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

using blast_demo::ScenePack;
using blast_demo::loadScenePack;

namespace
{

void require(bool condition, const std::string& message)
{
    if (!condition)
    {
        throw std::runtime_error("scene pack conformance failed: " + message);
    }
}

template <typename T>
void requireEqual(const T& actual, const T& expected, const std::string& field)
{
    if (!(actual == expected))
    {
        std::ostringstream out;
        out << field << ": got " << actual << ", golden says " << expected;
        throw std::runtime_error("scene pack conformance failed: " + out.str());
    }
}

void requireNear(double actual, double expected, double tolerance, const std::string& field)
{
    if (std::fabs(actual - expected) > tolerance)
    {
        std::ostringstream out;
        out << field << ": got " << actual << ", golden says " << expected;
        throw std::runtime_error("scene pack conformance failed: " + out.str());
    }
}

} // namespace

int main(int argc, char** argv)
{
    if (argc < 2)
    {
        std::fprintf(stderr, "usage: scene_pack_conformance_test <fixture.json>\n");
        return 2;
    }

    try
    {
        const ScenePack pack = loadScenePack(argv[1]);

        // --- Golden digest, mirrored in the TS and Rust conformance tests ---
        requireEqual<std::uint32_t>(pack.version, 2u, "version");
        requireEqual<std::size_t>(pack.nodes.size(), 7u, "nodeCount");
        requireEqual<std::size_t>(pack.bonds.size(), 7u, "bondCount");
        requireEqual<std::size_t>(pack.materials.size(), 3u, "materialCount");

        const std::vector<std::string> expectedNames{
            "reinforced-concrete", "concrete", "drywall-track"};
        for (std::size_t i = 0; i < expectedNames.size(); ++i)
        {
            requireEqual(pack.materials[i].name, expectedNames[i],
                         "materialNames[" + std::to_string(i) + "]");
        }

        std::size_t supportNodes = 0;
        double totalMass = 0.0;
        for (const auto& node : pack.nodes)
        {
            if (node.mass == 0.0f) ++supportNodes;
            totalMass += node.mass;
        }
        requireEqual<std::size_t>(supportNodes, 2u, "supportNodeCount");
        requireNear(totalMass, 6000.0, 1e-3, "totalMassKg");

        double totalArea = 0.0;
        std::vector<std::size_t> bondsPerMaterial(pack.materials.size(), 0);
        std::map<std::string, std::size_t> bondsPerClass;
        for (const auto& bond : pack.bonds)
        {
            totalArea += bond.area;
            require(bond.material < bondsPerMaterial.size(), "bond material index in range");
            ++bondsPerMaterial[bond.material];
            std::string a = pack.nodeTypes[bond.node0];
            std::string b = pack.nodeTypes[bond.node1];
            if (b < a) std::swap(a, b);
            ++bondsPerClass[a + "~" + b];
        }
        requireNear(totalArea, 1.46, 1e-6, "totalBondAreaM2");

        const std::vector<std::size_t> expectedPerMaterial{2, 2, 3};
        for (std::size_t i = 0; i < expectedPerMaterial.size(); ++i)
        {
            requireEqual(bondsPerMaterial[i], expectedPerMaterial[i],
                         "bondsPerMaterial[" + std::to_string(i) + "]");
        }

        const std::map<std::string, std::size_t> expectedPerClass{
            {"column~foundation", 2}, {"column~infill", 2},
            {"column~slab", 2}, {"infill~infill", 1}};
        requireEqual(bondsPerClass.size(), expectedPerClass.size(), "jointClassCount");
        for (const auto& entry : expectedPerClass)
        {
            const auto found = bondsPerClass.find(entry.first);
            require(found != bondsPerClass.end(), "missing joint class " + entry.first);
            requireEqual(found->second, entry.second, "bondsPerJointClass[" + entry.first + "]");
        }

        requireNear(pack.gravity, -9.81, 1e-6, "gravity");
        requireNear(pack.contactForceScale, 1.0, 1e-6, "contactForceScale");

        // The material index must reach the right bonds, not just count right:
        // the facade clips are drywall-track, the footings the frame default.
        requireEqual(pack.materials[pack.bonds[0].material].name,
                     std::string("reinforced-concrete"), "footing bond material");
        requireEqual(pack.materials[pack.bonds[4].material].name,
                     std::string("drywall-track"), "facade clip material");

        // Bonds 0 and 1 omit `m` entirely — absent must mean 0, not "unset".
        requireEqual<std::uint32_t>(pack.bonds[0].material, 0u, "omitted m defaults to 0");
    }
    catch (const std::exception& error)
    {
        std::fprintf(stderr, "%s\n", error.what());
        return 1;
    }

    std::printf("scene pack conformance (C++) passed\n");
    return 0;
}
